media-placeholders/src/main/java/org/wordpress/aztec/placeholders/PlaceholderManager.kt

```diff
@@ -0,0 +1,338 @@
+        get() = Dispatchers.Main + job
+    private val drawers = mutableMapOf<String, PlaceholderAdapter>()
+    private val positionToId = mutableSetOf<Placeholder>()
+
+    init {
+        aztecText.contentChangeWatcher.registerObserver(this)
+    }
+
+    fun onDestroy() {
+        aztecText.contentChangeWatcher.unregisterObserver(this)
+        job.cancel()
+        drawers.clear()
+    }
+
+    /**
+     * Register a custom drawer to draw a custom view over a placeholder.
+     */
+    fun registerDrawer(placeholderAdapter: PlaceholderAdapter) {
+        drawers[placeholderAdapter.type] = placeholderAdapter
+    }
+
+    /**
+     * Call this method to manually insert a new item into the aztec text. There has to be a drawer associated with the
+     * item type.
+     * @param id placeholder ID
+     * @param type placeholder type
+     * @param attributes other attributes passed to the view. For example a `src` for an image.
+     */
+    fun insertItem(id: String, type: String, vararg attributes: Pair<String, String>) {
+        if (drawers[type] == null) throw IllegalArgumentException("Drawer for inserted type not found. Register it with `registerDrawer` method")
+        val attrs = getAttributesForMedia(id, type, attributes)
+        val drawer = drawers[type] ?: return
+        val drawable = buildPlaceholderDrawable(drawer)
+        aztecText.insertMediaSpan(AztecPlaceholderSpan(aztecText.context, drawable, 0, attrs,
+                this, aztecText))
+        insertContentOverSpanWithId(attrs.getValue(UUID_ATTRIBUTE), null)
+    }
+
+    private fun buildPlaceholderDrawable(adapter: PlaceholderAdapter): Drawable {
+        val drawable = ContextCompat.getDrawable(aztecText.context, android.R.color.transparent)!!
+        drawable.setBounds(0, 0, aztecText.maxImagesWidth, adapter.getHeight(aztecText.maxImagesWidth))
+        return drawable
+    }
+
+    private fun updateAllBelowSelection(selectionStart: Int) {
+        positionToId.filter {
+            it.elementPosition >= selectionStart - 1
+        }.forEach {
+            insertContentOverSpanWithId(it.uuid, it.elementPosition)
+        }
+    }
+
+    private fun insertContentOverSpanWithId(uuid: String, currentPosition: Int? = null) {
+        var aztecAttributes: AztecAttributes? = null
+        val predicate = object : AztecText.AttributePredicate {
+            override fun matches(attrs: Attributes): Boolean {
+                val match = attrs.getValue(UUID_ATTRIBUTE) == uuid
+                if (match) {
+                    aztecAttributes = attrs as AztecAttributes
+                }
+                return match
+            }
+        }
+        val targetPosition = aztecText.getElementPosition(predicate) ?: return
+
+        insertInPosition(aztecAttributes ?: return, targetPosition, currentPosition)
+    }
+
+    private fun insertInPosition(attrs: AztecAttributes, targetPosition: Int, currentPosition: Int? = null) {
+        validateAttributes(attrs)
```

Comment:
This check doesn't currently do anything if it fails. Maybe do something like 
```suggestion
        if (!validateAttributes(attrs)) {
             // throw something?
        }
```